src/webxdc/integration.rs

```diff
@@ -0,0 +1,131 @@
+use crate::chat::{send_msg, ChatId};
+use crate::config::Config;
+use crate::contact::ContactId;
+use crate::context::Context;
+use crate::message::{Message, MsgId, Viewtype};
+use crate::param::Param;
+use crate::webxdc::{maps_integration, StatusUpdateItem, StatusUpdateSerial};
+use anyhow::Result;
+
+impl Context {
+    /// Set Webxdc file as integration.
+    /// @param file The .xdc file to use as Webxdc integration.
+    pub async fn set_webxdc_integration(&self, file: &str) -> Result<()> {
+        let chat_id = ChatId::create_for_contact(self, ContactId::SELF).await?;
+        let mut msg = Message::new(Viewtype::Webxdc);
+        msg.set_file(file, None);
+        msg.hidden = true;
+        msg.param.set_int(Param::WebxdcIntegration, 1);
+        msg.param.set_int(Param::GuaranteeE2ee, 1); // needed to pass `internet_access` requirements
+        send_msg(self, chat_id, &mut msg).await?;
+        Ok(())
+    }
+
+    /// Get Webxdc instance used for optional integrations.
+    /// @param integrate_for The chat to get the integration for.
+    /// @return Message ID that refers to the Webxdc instance; UI can open the Webxdc as usual.
+    ///     `None` if there is no integration; the caller can add one using `set_webxdc_integration` then.
+    pub async fn init_webxdc_integration(
+        &self,
+        integrate_for: Option<ChatId>,
+    ) -> Result<Option<MsgId>> {
+        let Some(instance_id) = self
+            .get_config_parsed::<u32>(Config::WebxdcIntegration)
+            .await?
+        else {
+            return Ok(None);
+        };
+
+        let Some(mut instance) =
+            Message::load_from_db_optional(self, MsgId::new(instance_id)).await?
+        else {
+            return Ok(None);
+        };
+
+        if instance.viewtype != Viewtype::Webxdc {
+            return Ok(None);
+        }
+
+        let integrate_for = integrate_for.unwrap_or_default().to_u32() as i32;
+        if instance.param.get_int(Param::WebxdcIntegrateFor) != Some(integrate_for) {
+            instance
+                .param
+                .set_int(Param::WebxdcIntegrateFor, integrate_for);
+            instance.update_param(self).await?;
+        }
+        Ok(Some(instance.id))
+    }
+
+    // Check if a Webxdc shall be used as an integration and remember that.
+    pub(crate) async fn update_webxdc_integration_database(&self, msg: &Message) -> Result<()> {
+        if msg.viewtype == Viewtype::Webxdc && msg.param.get_int(Param::WebxdcIntegration).is_some()
+        {
+            self.set_config_internal(
```

Comment:
Won't it be more straightforward to do this in `set_webxdc_integration()` after the `send_msg()` call?

Reply:
this does not work as `send_msg()` already needs to know that it is an integration (integrations  are not sent over the wire).

this might be refactored, but that seems over the top, esp. as there is also the rough idea of letting the user/dev add/overwrite integrations on their own by adding to "Saved Messages", so, `set_webxdc_integration()` is not even called in  that case.